stacks/spica/src/passport/directives/can-interact.directive.spec.ts

```diff
@@ -0,0 +1,81 @@
+import {Component, ViewChild} from "@angular/core";
+import {TestBed, ComponentFixture} from "@angular/core/testing";
+import {CanInteractDirective} from "./can-interact.directive";
+import {PassportService} from "../services/passport.service";
+import {of} from "rxjs";
+
+@Component({
+  template: `
+    <button [canInteract]="condition">Click here.</button>
+  `
+})
+class TestComponent {
+  @ViewChild(CanInteractDirective, {static: true}) directive: CanInteractDirective;
+  condition = false;
+}
+
+class TestPassportService {
+  checkAllowed = (value: boolean) => {
+    return of(value);
+  };
+}
+
+describe("CanInteract", () => {
+  let component: TestComponent;
+  let fixture: ComponentFixture<TestComponent>;
+  let button: HTMLElement;
+  let setVisibleSpy: jasmine.Spy;
+
+  beforeEach(async () => {
+    TestBed.configureTestingModule({
+      declarations: [TestComponent, CanInteractDirective],
+      providers: [{provide: PassportService, useClass: TestPassportService}]
+    });
+
+    fixture = TestBed.createComponent(TestComponent);
+    component = fixture.componentInstance;
+
+    setVisibleSpy = spyOn(component.directive, "setVisible").and.callThrough();
+
+    fixture.detectChanges();
+
+    //wait until setting initial value from passport service
+    await fixture.whenStable();
+    fixture.detectChanges();
+
+    button = fixture.debugElement.nativeElement.querySelector("button");
+  });
+
+  it("shouldn't show button if condition is false", () => {
+    expect(setVisibleSpy).toHaveBeenCalledTimes(1);
+    expect(setVisibleSpy).toHaveBeenCalledWith(false);
+    expect(button.style.visibility).toEqual("hidden");
+  });
+
+  it("should change visible property of button when condition changed", async () => {
+    setVisibleSpy.calls.reset();
```

Comment:
Why do you reset calls before at the beginning? jasmine does it for you already. As you can see you create a new spy in beforeEach calls.

Reply:
All beforeEaches create component and pass the condition value to canInteract directive. It means when the fixture is ready right after component created, canInteract calls this setVisible method inside of ngOnInit method. So, when the **it** blocks ready to start, spy have been called 1 time already. If I don't reset these calls, I have to write `toHaveBeenCalledTimes(2)` cause of the initial call that I explained. If I create this spy inside of these **it** blocks, there will be no chance to test the initial spy call.